Implement *Assign for BigUint

Darksonn opened the pull request:
Not only does this change increase convenience of use, it also allows adding a `&BigUint` to a `&mut BigUint` without allocating (if not necessary) or tricks such as:

    fn add(a: &mut BigUint, b: &BigUint) {
        let aa = mem::replace(a, BigUint::from_slice(&[])); // BigUint::from_slice(&[]) does not allocate
        *a = aa + b;
    }

With this change:

    fn add(a: &mut BigUint, b: &BigUint) {
        *a += b;
    }

It would make sense to add the same functionality to `BigInt`, but it uses some macros to handle the signs, and I'm not sure how to change the macros in order to perform this change.

cuviper commented:
Thanks!  The implementation looks good, but I'd like a few more extensions to make this complete.

The binary ops all use forwarding to resolve ref/value variants, but you didn't do this for assignment ops.  e.g. you have 'impl<'a> AddAssign<&'a BigUint> for BigUint`, but not `AddAssign<BigUint>` which could just forward to the other.

There are more assignment ops available -- is there any reason you only did these four?  At a minimum, I'd also like `RemAssign`.  In combination with by-value assign ops, it would complete `NumAssign` and `NumAssignRef`.  For extra credit, we could also have the shifting and bitwise assign ops (`ShrAssign`, BitAndAssign`, etc.)

On the scalar side, there are the `promote_scalars` macros for additional sizes, which are missing here for assignment ops.

I know, it's a mess of macros creating a ton of implementations, but I think they're useful, and it would be nice for these new additions to be equally broad.

(And I agree `BigInt` would be nice too, but we can treat that separately after this goes in.)

Darksonn commented:
I found out how your macros work, so I've implemented by-value assign. I also added the remaining operators and some tests.

I implemented `scalar %= BigUint` for all the scalars as well; it makes sense since the result will always fit in the scalar. The same could be done for BitAnd, Shr and Shl, but I haven't added those, since you didn't even have the standard BitAnd, Shr and Shl operators for scalars. I didn't add a test for `scalar %= BigUint`, since I wasn't sure of a nice way to test it.

It's also a bit weird that `BigUint` implements `Neg`, though that's not really relevant to this.

cuviper commented:
I made a few small tweaks for CI and such, but otherwise this is great!  And wow, that's a lot of implementations on `BigUint` now.  Let's merge it!

bors r+
